File: services/client/src/lottery/protocol/packets/packet.go
package packets

import (
	"errors"
	"math"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/lottery"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/lottery/protocol/packets/bet_packet"
)

const MaxBets = math.MaxUint8

type Packet struct {
	header Header
	bets   []bet_packet.BetPacket
}

func newHeaderPacket(header Header) *Packet {
	bets := make([]bet_packet.BetPacket, 0)

	return &Packet{header, bets}
}

func NewAckPacket(agencyId uint8) *Packet {
	return newHeaderPacket(NewAckHeader(agencyId))
}

func NewBetPacket(agencyId uint8, bets []lottery.Bet) (*Packet, error) {
	betAmount := uint8(len(bets))

	if betAmount == 0 {
		return nil, errors.New("bets is empty")
	}
	if betAmount > MaxBets {
		return nil, errors.New("too many bets")
	}
	betPackets := make([]bet_packet.BetPacket, betAmount)

	var betSize uint16 = 0

	for i, bet := range bets {
		betPacket, err := bet_packet.NewBetPacket(bet)

		if err != nil {
			return nil, err
		}
		betPackets[i] = *betPacket
		betSize += betPacket.Size()
	}
	header := NewBetHeader(betAmount, betSize, agencyId)
	return &Packet{header, betPackets}, nil
}

func NewEndPacket(agencyId uint8) *Packet {
	return newHeaderPacket(NewEndHeader(agencyId))
}

func (packet *Packet) ToBytes() ([]byte, error) {
	bytes := make([]byte, HeaderSize+packet.header.BetSize)

	packet.header.WriteToBytes(bytes)

	var i uint16 = HeaderSize

	for _, bet := range packet.bets {
		bet.WriteToBytes(bytes[i:])
		i += bet.Size()
	}
	return bytes, nil
}

func AddBetsFromBytes(header *Header, bytes []byte, bets []lottery.Bet) error {
	var i uint16

	for i < uint16(header.BetAmount) {
		packet, err := bet_packet.BetPacketFromBytes(bytes[i:])

		if err != nil {
			return err
		}
		bets = append(bets, packet.ToBet())
		i += packet.Size()
	}
	return nil
}

File: services/server/src/lottery/protocol/packets/packet.py
from typing import List

from lottery import Bet
from lottery.protocol.packets.bet_packet.bet_packet import BetPacket
from lottery.protocol.packets.header import Header

MAX_BETS = 255


class Packet:
    def __init__(self, header: Header, bets: List[BetPacket] = []):
        self.header = header
        self.bets = bets

    @classmethod
    def new_ack_packet(cls, agency_id: int) -> 'Packet':
        return cls(Header.new_ack_header(agency_id))

    @classmethod
    def new_end_packet(cls, agency_id: int) -> 'Packet':
        return cls(Header.new_end_header(agency_id))

    @classmethod
    def new_bet_packet(cls, agency_id: int, bets: List[Bet]) -> 'Packet':
        bet_amount = len(bets)

        if bet_amount == 0:
            raise ValueError('bets is empty')

        if bet_amount > MAX_BETS:
            raise ValueError('too many bets')

        bet_packets = []
        bet_size = 0

        for bet in bets:
            bet_packet = BetPacket.from_bet(bet)
            bet_packets.append(bet_packet)
            bet_size += bet_packet.size()

        header = Header.new_bet_header(bet_amount, bet_size, agency_id)

        return cls(header, bet_packets)

    def to_bytes(self) -> bytearray:
        bytes = bytearray()

        self.header.write_to_bytes(bytes)

        for bet in self.bets:
            bet.write_to_bytes(bytes)

        return bytes

    @classmethod
    def add_bets_from_bytes(cls, header: Header, bytes: bytes, bets: List[Bet]) -> None:
        offset = 0

        while offset < header.bet_size:
            packet = BetPacket.from_bytes(bytes[offset:], header.agency_id)
            bets.append(packet.bet)
            offset += packet.size()
